### PR TITLE
NUnit assertion to fluentassertions migration refactorings. Patterns for Should().NotBeNull()

### PR DESCRIPTION
**Pull request by inyutin-maxim**
Added highlighting and quick fix for `Assert.IsNotNull(result)` and for `Assert.NotNull(result)`.

I would like to get feedback on the current implementation.

Also, a lot of new functionality is planned, I plan to send a separate PR for each feature.

If you want something different then please explain how you want to build an interaction.

Thank you

### CODE REVIEW

**Comment by matkoch** [comment]
Hey Maxim, I should've maybe mentioned that before, but have you thought about using SSR patterns for that? Those would require much less code. The only reason I can see manual implementation would be better is if it would support "bulk fixes".

**Comment by inyutin-maxim** [comment]
@matkoch I think "bulk fixes" is must valuable criterion (for me for example).
I will be able to extract some base classes for the generic code when I have a little more implementations. 
I will study the documentation about SSR, thank you for the offer. 
It does not scare me to write the required code, especially since I have no deadlines `:-)`

**Comment by inyutin-maxim** [comment]
Hi @matkoch, I think I have completed the architecture for this feature, and would like to get feedback from you, for example, that could be improved or rewritten otherwise so that I could scale the resulting result to other expressions

**Comment by matkoch** [comment]
Hey @inyutin-maxim. Thanks for your PR! I was on vacation and will follow up very shortly on this.

**Comment by inyutin-maxim** [comment]
Hi @matkoch, do you have any feedback?

**Comment by matkoch** [comment]
Hi @inyutin-maxim. First of all sorry for the late response. Didn't go as I wanted =)

Now some feedback:

File: src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/NUnit/AssertNotBeNullMigrationService.cs
Comment (matkoch) [inline src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/NUnit/AssertNotBeNullMigrationService.cs]: I haven't run the code yet, but this looks suspicious. Will it also use `NotBeNull` when converting from `Assert.Null` ?

File: src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/TestingFrameworkScanner.cs
Comment (matkoch) [inline src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/TestingFrameworkScanner.cs]: For all the logic that determines whether a project is referencing a certain assembly, or whether a call is actually an NUnit call, you can [take a look here](https://github.com/matkoch/resharper-fluentvalidation/tree/master/src/dotnet/ReSharperPlugin.FluentValidation/Psi). It should be mostly pattern matching (not the syntax thing :)) with other namespaces and type fullnames, if you have any questions, feel free to ask.

File: src/dotnet/ReSharperPlugin.FluentAssertions/QuickFixes/NUnitAssertMigrationQuickFix.cs
Comment (matkoch) [inline src/dotnet/ReSharperPlugin.FluentAssertions/QuickFixes/NUnitAssertMigrationQuickFix.cs]: Might be just personal preference, but I wouldn't make it so complicated by having an interface etc..

A better way might be having a base problem analyzer, base highlighting, and base quickfix. Most of the NUnit logic is then in the base class, and the derived class can make UI texts more specific, for instance like `Replace Assert.Null with obj.Should().BeNull()`

**Review by matkoch — CHANGES_REQUESTED** [review]
Hi @inyutin-maxim. First of all sorry for the late response. Didn't go as I wanted =)
Kudos for also providing the test infrastructure!

I left some review comments. Let me know if you need help!

File: src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/NUnit/AssertNotBeNullMigrationService.cs
Comment (inyutin-maxim) [reply src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/NUnit/AssertNotBeNullMigrationService.cs]: Not of course. 
This service implements interface [INUnitAssertMigrationService](https://github.com/inyutin-maxim/resharper-fluentassertions/blob/1c1e2fc7844164078c4569d3984b3c867163e2f2/src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/NUnit/INUnitAssertMigrationService.cs#L9)

Interface `INUnitAssertMigrationService` contain method [CanMigrate](https://github.com/inyutin-maxim/resharper-fluentassertions/blob/master/src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/NUnit/AssertNotBeNullMigrationService.cs#L34-L38) where i check invokated method name.
For all concrete methods, I create new implementations for `INUnitAssertMigrationService` which create concrete migration expression.

Examples:
- For `Assert.NotNull` or `Assert.IsNotNull` i have [AssertNotBeNullMigrationService](https://github.com/inyutin-maxim/resharper-fluentassertions/blob/52837ea6b45ef015751718c17e9603093e909bae/src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/NUnit/AssertNotBeNullMigrationService.cs#L27) which create `result.Should().NotBeNull()";`
- For `Assert.IsNotEmpty` i have [AssertIsNotEmptyMigrationService](https://github.com/inyutin-maxim/resharper-fluentassertions/blob/52837ea6b45ef015751718c17e9603093e909bae/src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/NUnit/AssertIsNotEmptyMigrationService.cs#L28) which create `result.Should().NotBeEmpty()";`
- etc

File: src/dotnet/ReSharperPlugin.FluentAssertions/QuickFixes/NUnitAssertMigrationQuickFix.cs
Comment (inyutin-maxim) [reply src/dotnet/ReSharperPlugin.FluentAssertions/QuickFixes/NUnitAssertMigrationQuickFix.cs]: `Assert.Null`, `Assert.NotNull`, `Assert.IsNotEmpty` - this simple invocations, they can implement base class, but NUnit Assert contains more complex methods for example: `Assert.That`, `Assert.Throws`
They contain much more complex logic.
Therefore, the interface suits me more, but maybe these are really personal preferences :-)
I also thought about more specific messages, I add this it is a little later

File: src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/TestingFrameworkScanner.cs
Comment (inyutin-maxim) [reply src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/TestingFrameworkScanner.cs]: Thanks for the examples, I will study and rewrite this service. I don't quite like my implementation either, but it works `:-)`

File: src/dotnet/ReSharperPlugin.FluentAssertions/QuickFixes/NUnitAssertMigrationQuickFix.cs
Comment (inyutin-maxim) [reply src/dotnet/ReSharperPlugin.FluentAssertions/QuickFixes/NUnitAssertMigrationQuickFix.cs]: I have a question. 
How I can create InvocationExpression for unit testing, without all testing infrastructure for integration testing?
If it is possible of course :-)

**Comment by inyutin-maxim** [comment]
Hi @matkoch, when I start plugin with `Rider (Windows)` configuration, rider throw exception:

```
Message = “Не удалось загрузить файл или сборку "nunit.framework, Version=3.12.0.0, Culture=neutral, PublicKeyToken=2638cd05610744eb" либо одну из их зависимостей. Не удается найти указанный файл.”
ExceptionPath = Root.InnerException
ClassName = System.IO.FileNotFoundException
HResult = COR_E_FILENOTFOUND=E_FILENOTFOUND=WIN32_ERROR_FILE_NOT_FOUND=80070002
Source = ReSharperPlugin.FluentAssertions
FileNotFound_FileName = “nunit.framework, Version=3.12.0.0, Culture=neutral, PublicKeyToken=2638cd05610744eb”
FileNotFound_FusionLog = “
  Предупреждение: регистрация привязки сборок выключена.
  Чтобы включить регистрацию ошибок привязки сборок, установите значение параметра реестра [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) в 1.
  Примечание. Регистрация ошибок привязки сборок может привести к некоторому снижению производительности.
  Чтобы отключить эту функцию, удалите параметр реестра [HKLM\Software\Microsoft\Fusion!EnableLog].
”
StackTraceString = “
  в ReSharperPlugin.FluentAssertions.Analyzers.NUnitAssertMigrationAnalyzer.Run(IInvocationExpression element, ElementProblemAnalyzerData data, IHighlightingConsumer consumer)
     в JetBrains.ReSharper.Feature.Services.Daemon.ElementProblemAnalyzerRegistrar.ElementAnalyzerDispatcher.Run(ITreeNode element, IHighlightingConsumer consumer)

```

I add the NuGet package to the project with plugin:
```
 <ItemGroup>
    <PackageReference Include="NUnit" Version="3.12.0" PrivateAssets="all" />
    <PackageReference Include="FluentAssertions" Version="5.10.3" PrivateAssets="all" />
    <PackageReference Include="JetBrains.ReSharper.SDK" Version="$(SdkVersion)" PrivateAssets="all" />
    <!-- TODO: https://github.com/NuGet/Home/issues/7154 -->
    <PackageReference Include="Wave" Version="[$(WaveVersion),$(UpperWaveVersion))" />
  </ItemGroup>
```

I thought it would help me, but obviously, the problem is something else. If you know how to fix the problem then it would be great if you helped with this, and for now, I will continue to find a solution myself

File: src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/TestingFrameworkScanner.cs
Comment (inyutin-maxim) [reply src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/TestingFrameworkScanner.cs]: done with https://github.com/inyutin-maxim/resharper-fluentassertions/commit/29533d10c900169579dbfb4c9d973bb66d4461bc

**Comment by inyutin-maxim** [comment]
I have a question.
How I can create InvocationExpression for unit testing, without all testing infrastructure for integration testing?
If it is possible of course :-)

File: src/dotnet/ReSharperPlugin.FluentAssertions/QuickFixes/NUnitAssertMigrationQuickFix.cs
Comment (inyutin-maxim) [reply src/dotnet/ReSharperPlugin.FluentAssertions/QuickFixes/NUnitAssertMigrationQuickFix.cs]: I fixed your remark in commit https://github.com/matkoch/resharper-fluentassertions/pull/6/commits/cc0710f3739159b0d5f06343f9df808462e61df2

**Comment by inyutin-maxim** [comment]
Also, I have a method that determines whether an `IsMemberAccessExpressionTypeOf<T>`  what do you think about implementation?

For example, I want to be sure that Assert belongs to NUnit and not xUnit
``` csharp
string nullString = null;
Assert.IsNull(nullString);
^^^^^^
```


https://github.com/inyutin-maxim/resharper-fluentassertions/blob/master/src/dotnet/ReSharperPlugin.FluentAssertions/Analyzers/NUnitAssertMigrationAnalyzer.cs#L42-L68

**Comment by matkoch** [comment]
Theoretically, you can reference other libraries in your plugin, but I wouldn't exactly recommend it. If you really want to go that way, I have to ask around how these additional assemblies can be loaded in tests. Let me know before I go ahead :)

**Comment by matkoch** [comment]
Regarding the check if `IsNull` originates from NUnit, the better approach here is to use `Resolve` on the invocation expression, and check the declaring type from the returned `IDeclaredElement`.

**Comment by inyutin-maxim** [comment]
> Regarding the check if `IsNull` originates from NUnit, the better approach here is to use `Resolve` on the invocation expression, and check the declaring type from the returned `IDeclaredElement`.

Thanks, I try implement check with you recommendations

**Comment by inyutin-maxim** [comment]
> Theoretically, you can reference other libraries in your plugin, but I wouldn't exactly recommend it. If you really want to go that way, I have to ask around how these additional assemblies can be loaded in tests. Let me know before I go ahead :)

If I am correct understand you, that for tests i use attribute `[TestNet50("NUnit/3.12.0", "FluentAssertions/5.10.3")]`
where I can set concrete dependency packages.

But my question is that in an assembly with the plugin itself, the constraint "nunit.framework, Version = 3.13.2.0, Culture = neutral, PublicKeyToken=2638cd05610744eb" is not loaded at start.

Best way to exclude all references on NUnit from a plugin project and use string constants instead?

I want to use reference to NUnit for use for example `nameof(Assert.NotNull)`

**Comment by inyutin-maxim** [comment]
Hi, @matkoch I have one more question.
Now the plugin works as I expect with the exception of one nuance. The patch works only for a single invocation but does not apply to the entire file or project, or solution

**Comment by inyutin-maxim** [comment]
> Hi, @matkoch I have one more question.
> Now the plugin works as I expect with the exception of one nuance. The patch works only for a single invocation but does not apply to the entire file or project, or solution

Fixed in the last commit

**Comment by inyutin-maxim** [comment]
Hi @matkoch, I think I have completed my fixes, if you have any other comments or suggestions, I am ready to correct them

**Comment by matkoch** [comment]
@inyutin-maxim Just a heads up, I will get to it this or coming week. Sorry, the past week and this are full of release stuff. But I'm really excited that you're so committed :)

**Comment by inyutin-maxim** [comment]
Hi @matkoch, I hope I have not been forgotten, I am still waiting for feedback from you on the implementation to continue working on the rest of the transformations based on the already formed architecture for this feature :-)

**Comment by inyutin-maxim** [comment]
> Hi @matkoch, I hope I have not been forgotten, I am still waiting for feedback from you on the implementation to continue working on the rest of the transformations based on the already formed architecture for this feature :-)

Hi @matkoch, I hope I have not been forgotten. I wait for feedback from you. :-)

**Comment by matkoch** [comment]
Hey @inyutin-maxim! I'm terribly sorry, it keeps dropping on my list because of other urgent things :/ One thing though: are you on our plugin developers Slack channel? If not I could invite you, and it would be much easier to give feedback via voice calls. Written feedback takes much longer, and I wouldn't be mad if you ping me daily :)

**Comment by inyutin-maxim** [comment]
> Hey @inyutin-maxim! I'm terribly sorry, it keeps dropping on my list because of other urgent things :/ One thing though: are you on our plugin developers Slack channel? If not I could invite you, and it would be much easier to give feedback via voice calls. Written feedback takes much longer, and I wouldn't be mad if you ping me daily :)

Hi, @matkoch how I can join your plugin developer's Slack channel?
my email: inyutin.maxim@gmail.com

**Comment by matkoch** [comment]
You should receive an invitation latest on Monday.

File: src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/ExpressionFormatParametersExtensions.cs
Comment (ForNeVeR) [inline src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/ExpressionFormatParametersExtensions.cs]: _(Minor)_ This could theoretically break on systems with weird locales. I'd recommend to always use `InvariantCulture` when formatting any kind of number.

File: src/dotnet/ReSharperPlugin.FluentAssertions/Psi/FluentAssertionsPredefinedType.cs
Comment (ForNeVeR) [inline src/dotnet/ReSharperPlugin.FluentAssertions/Psi/FluentAssertionsPredefinedType.cs]: To be honest, I don't get why do we need this complicated system which stores indices of static fields in a dictionary.

**Review by ForNeVeR — CHANGES_REQUESTED** [review]

File: src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/ExpressionFormatParametersExtensions.cs
Comment (inyutin-maxim) [reply src/dotnet/ReSharperPlugin.FluentAssertions/Helpers/ExpressionFormatParametersExtensions.cs]: 9e71b8b

**Comment by inyutin-maxim** [comment]
Hi @matkoch I completed the corrections on your remarks, unfortunately, I ruined the story a little and it duplicated after merging changes. If you want I can create a new request to another branch, the feature is not yet complete and covers the fix for only one case, but discovers all situations when invoked NUnit.Assert. What do you think of that?

File: src/dotnet/ReSharperPlugin.FluentAssertions/Highlightings/NUnitAssertMigrationHighlighting.cs
Comment (matkoch) [inline src/dotnet/ReSharperPlugin.FluentAssertions/Highlightings/NUnitAssertMigrationHighlighting.cs]: nit: `FluentAssertions`

**Comment by matkoch** [comment]
I didn't run tests or anything, but LGTM I think :) Can also merge without the nit fix.

**Comment by matkoch** [comment]
Are you good with squash-merge ?

File: src/dotnet/ReSharperPlugin.FluentAssertions/Highlightings/NUnitAssertMigrationHighlighting.cs
Comment (inyutin-maxim) [reply src/dotnet/ReSharperPlugin.FluentAssertions/Highlightings/NUnitAssertMigrationHighlighting.cs]: fixed in last commit

**Comment by inyutin-maxim** [comment]
> I didn't run tests or anything, but LGTM I think :) Can also merge without the nit fix.

image

**Comment by inyutin-maxim** [comment]
> Are you good with squash-merge ?

Unfortunately not good enough in this, I can certainly try in myself in a separate branch, but so far I do not imagine how

**Comment by inyutin-maxim** [comment]
Hi, @matkoch do you have any other comments? if not, would be happy to continue working on the following quick fixes, when you merge my PR

**Pull request merged**
